tainting: Do not analyze named function definitions twice

IagoAbal opened the pull request:
We were visiting named function definitions twice, one time with `kdef`,
and a second time with `kfunction_definition`!

Closes PA-1778

test plan:
% semgrep --metrics=off -c p/java lang/java/tmp/elastic-elasticsearch
  #^ This is now ~1.3x faster

PR checklist:

- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] `changelog.d/<issue>.<type>` is a file with the _what_, _why_, and _how_ of the change.
  - \<issue> is `pa-312` (Linear ticket), `gh-1234` (GitHub issue), or `new-gizmo` (unique semantic name)
  - \<type> is `added`, `changed`, `fixed`, or `infra`.
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Changelog guidelines](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry)
- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

IagoAbal committed:
tainting: Do not analyze named function definitions twice

We were visiting named function definitions twice, one time with `kdef`,
and a second time with `kfunction_definition`!

Closes PA-1778

test plan:
% semgrep --metrics=off -c p/java lang/java/tmp/elastic-elasticsearch
  #^ This is now ~1.3x faster

aryx commented on semgrep-core/src/core/ast/Visitor_AST.ml:
```diff
@@ -100,9 +100,11 @@ let (mk_visitor :
       visitor_out) =
- fun ?(vardef_assign = false) ?(flddef_assign = false) ?(attr_expr = false) vin ->
+ fun ?(vardef_assign = false) ?(flddef_assign = false) ?(attr_expr = false)
+     ?(fundef_as_anon = true) vin ->
```
hmm all the default value are set to false, so maybe we should make this more consistent and set it to false too?

IagoAbal replied:
I agree, although it was a bit more cumbersome to make it `false` by default since one needs to review every use of `mk_visitor` to check where we should enable this (which was the default before). So this made me reconsider the solution, and I think the one I pushed now it's simpler and still seems to give good perf.

IagoAbal committed:
Yoann's review

Yoann's request to have `fundef_as_anon` disabled by default made me
reconsider the solution entirely. I ran the test plan and got the
same result.
